### PR TITLE
[Don't merge]support Mutiview Matrix

### PR DESCRIPTION
**Pull request by songchengjiang**
In order to support GLES Mutiview extension

### CODE REVIEW

File: cocos/2d/CCScene.h [hunk 1]
```diff
@@ -109,7 +109,7 @@ class CC_DLL Scene : public Node
      * @param renderer The renderer use to render the scene.
      * @js NA
      */
-    virtual void render(Renderer* renderer, const Mat4& eyeTransform, const Mat4* eyeProjection = nullptr);
+    virtual void render(Renderer* renderer, const Mat4* eyeTransforms = nullptr, const Mat4* eyeProjections = nullptr, unsigned int mutiViewCount = 1);
```
Comment (minggo) [inline cocos/2d/CCScene.h:112]: It is better to add a new function 

``` c++
virtual void render(Renderer* renderer, const Mat4* eyeTransforms = nullptr, const Mat4* eyeProjections = nullptr, unsigned int multiViewCount)
```

And there is an typo, `mutiViewCount` -> `multiViewCount`

File: cocos/base/CCDirector.h [hunk 2]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
      * Clones a specified type matrix and put it to the top of specified type of matrix stack.
      * @js NA
      */
-    void pushMatrix(MATRIX_STACK_TYPE type);
+    void pushMatrix(MATRIX_STACK_TYPE type, unsigned int index = 0);
```
Comment (minggo) [inline cocos/base/CCDirector.h:458]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 3]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
...
     /** Pops the top matrix of the specified type of matrix stack.
      * @js NA
      */
-    void popMatrix(MATRIX_STACK_TYPE type);
+    void popMatrix(MATRIX_STACK_TYPE type, unsigned int index = 0);
```
Comment (minggo) [inline cocos/base/CCDirector.h:462]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 4]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
...
     /** Adds an identity matrix to the top of specified type of matrix stack.
      * @js NA
      */
-    void loadIdentityMatrix(MATRIX_STACK_TYPE type);
+    void loadIdentityMatrix(MATRIX_STACK_TYPE type, unsigned int index = 0);
```
Comment (minggo) [inline cocos/base/CCDirector.h:466]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 5]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
...
     /**
      * Adds a matrix to the top of specified type of matrix stack.
      * 
      * @param type Matrix type.
      * @param mat The matrix that to be added.
      * @js NA
      */
-    void loadMatrix(MATRIX_STACK_TYPE type, const Mat4& mat);
+    void loadMatrix(MATRIX_STACK_TYPE type, const Mat4& mat, unsigned int index = 0);
```
Comment (minggo) [inline cocos/base/CCDirector.h:474]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 6]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
...
     /**
      * Multiplies a matrix to the top of specified type of matrix stack.
      *
      * @param type Matrix type.
      * @param mat The matrix that to be multiplied.
      * @js NA
      */
-    void multiplyMatrix(MATRIX_STACK_TYPE type, const Mat4& mat);
+    void multiplyMatrix(MATRIX_STACK_TYPE type, const Mat4& mat, unsigned int index = 0);
```
Comment (minggo) [inline cocos/base/CCDirector.h:482]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 7]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
...
     /**
      * Gets the top matrix of specified type of matrix stack.
      * @js NA
      */
-    const Mat4& getMatrix(MATRIX_STACK_TYPE type) const;
+    const Mat4& getMatrix(MATRIX_STACK_TYPE type, unsigned int index = 0) const;
```
Comment (minggo) [inline cocos/base/CCDirector.h:487]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 8]
```diff
@@ -455,41 +455,42 @@ class CC_DLL Director : public Ref
...
     /**
      * Clear all types of matrix stack, and add identity matrix to these matrix stacks.
      * @js NA
      */
-    void resetMatrixStack();
+    void resetMatrixStack(unsigned int stackCount = 1);
```
Comment (minggo) [inline cocos/base/CCDirector.h:492]: It is better to add a new function.

File: cocos/base/CCDirector.h [hunk 9]
```diff
@@ -525,10 +526,10 @@ class CC_DLL Director : public Ref
     void initTextureCache();
     void destroyTextureCache();
 
-    void initMatrixStack();
+    void initMatrixStack(unsigned int stackCount);
```
Comment (minggo) [inline cocos/base/CCDirector.h:529]: It is better to add a new function.

File: cocos/renderer/CCGLProgram.cpp [hunk 10]
```diff
@@ -116,8 +116,10 @@ const char* GLProgram::SHADER_CAMERA_CLEAR = "ShaderCameraClear";
 // uniform names
 const char* GLProgram::UNIFORM_NAME_AMBIENT_COLOR = "CC_AmbientColor";
 const char* GLProgram::UNIFORM_NAME_P_MATRIX = "CC_PMatrix";
+const char* GLProgram::UNIFORM_NAME_MUTIVIEW_P_MATRIX = "CC_MutiViewPMatrix";
```
Comment (minggo) [inline cocos/renderer/CCGLProgram.cpp:119]: a typo
CC_MutiViewPMatrix -> CC_MultiViewPMatrix

Comment (minggo) [reply cocos/renderer/CCGLProgram.cpp:119]: UNIFORM_NAME_MUTIVIEW_P_MATRIX -> UNIFORM_NAME_MULTIVIEW_P_MATRIX

File: cocos/renderer/CCGLProgram.cpp [hunk 11]
```diff
@@ -116,8 +116,10 @@ const char* GLProgram::SHADER_CAMERA_CLEAR = "ShaderCameraClear";
...
 const char* GLProgram::UNIFORM_NAME_MV_MATRIX = "CC_MVMatrix";
 const char* GLProgram::UNIFORM_NAME_MVP_MATRIX  = "CC_MVPMatrix";
+const char* GLProgram::UNIFORM_NAME_MUTIVIEW_MVP_MATRIX  = "CC_MutiViewMVPMatrix";
```
Comment (minggo) [inline cocos/renderer/CCGLProgram.cpp:122]: CC_MutiViewMVPMatrix -> CC_MultiViewMVPMatrix
UNIFORM_NAME_MUTIVIEW_MVP_MATRIX -> UNIFORM_NAME_MULTIVIEW_MVP_MATRIX

**Comment by minggo** [comment]
@ricardoquesada this PR is added to support ARM Mali VR, because ARM Mali VR depends on multiple views. Could you please take a look if the modification is suitable or not without detail inspect. @songchengjiang will continue to polish the PR(such as fix typo, add new functions instead of use default parameter) if the modification is ok.
Thanks.

**Comment by ricardoquesada** [comment]
The PR good looks to me. My only concern is the "multi views" parameter. What's the maximum number of views ? 
If the answer is two, like "left eye" and right "eye", I would make it more explicit , like "left eye" stack, and "right eye" stack.

**Comment by songchengjiang** [comment]
@ricardoquesada As ARM said, The scene is rendered twice for each eye position, once with a wide field of view, and once with a narrower field of view in order to render the midpoint of the screen with a higher resolution than the rest.

**Comment by ricardoquesada** [comment]
@songchengjiang ok. adding "explicit" names for the stacks, or better documentation is needed.

Otherwise developers (including myself) won't understand why we have multiple Transform stacks.

But this code won't work correctly in legacy code (like the Box2d debug code) that calls `GLProgram::setUniformsForBuiltins();`
It will "work", but not correctly, since it will always use the stack number 0.

If that is not important, then probably we shouldn't have multiple stacks.

**Comment by minggo** [comment]
@ricardoquesada i am not sure if it is possible to add `explicit` names for stacks. Yep, the meaning of ARM VR is clear, but is there any other VR that needs multiple views too but have different meanings with each view? So i think have good documents is better.

**Pull request closed without merging**